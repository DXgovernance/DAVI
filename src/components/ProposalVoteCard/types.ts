import { BigNumber } from 'ethers';
import type { VoteData } from 'Modules/Guilds/Hooks/useVotingResults';
import { ERC20Guild } from 'types/contracts';
import { ProposalMetadata } from 'types/types.guilds';
import { Moment } from 'moment';
import { providers } from 'ethers';

export interface VoteConfirmationModalProps {
  isOpen: boolean;
  onDismiss: () => void;
  selectedAction?: string;
  onConfirm: () => void;
  votingPower?: number;
  currentVoteAmount?: number;
}

export interface Voter {
  avatar: string;
}

export interface hasVotingPowerProps {
  hasNoVotingPower: boolean;
  hasVotingPowerAtCurrentSnapshot: boolean;
}

interface VotingPower {
  percent: number;
  userVotingPower: BigNumber;
  atSnapshot: BigNumber;
  atCurrentSnapshot: BigNumber;
}

interface Proposal {
  id: string;
  metadata: ProposalMetadata;
  endTime: Moment;
  title?: string;
}

export interface ProposalVoteCardProps {
  voteData: VoteData;
  proposal: Proposal;
  timestamp: number;
  votingPower: VotingPower;
  contract: ERC20Guild;
  createTransaction: (
    summary: string,
    txFunction: () => Promise<providers.TransactionResponse>
  ) => void;
}

export interface ResultRowProps {
  isPercent: boolean;
  optionKey?: number;
  voteData: VoteData;
  proposalMetadata: ProposalMetadata;
}

export interface VoteChartProps {
  isPercent: boolean;
  voteData: VoteData;
}

export interface Voter {
  avatar: string;
}

export interface ConfirmVoteProposalProps {
  proposal: Proposal;
  contract: ERC20Guild;
  selectedAction: BigNumber;
  userVotingPower: BigNumber;
  createTransaction: (
    summary: string,
    txFunction: () => Promise<providers.TransactionResponse>,
    showModal?: boolean,
    cb?: () => void
  ) => void;
  cb?: () => void;
}

export interface VoteOnProposalProps {
  votingPowerAtProposalSnapshot: BigNumber;
  votingPowerAtProposalCurrentSnapshot: BigNumber;
}

export type VoteResultsProps = Omit<ResultRowProps, 'optionKey'>;
