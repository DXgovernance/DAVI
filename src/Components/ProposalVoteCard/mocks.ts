import moment from 'moment';
import { BigNumber } from 'ethers';

const mockBignumber = BigNumber.from(1000);

export const mockProposalVoteCardProps = {
  proposal: {
    id: '0x0',
    metadata: {
      description: '',
      voteOptions: [],
    },
    endTime: moment(),
  },
  voteData: {
    options: {},
    quorum: mockBignumber,
    totalLocked: mockBignumber,
    token: {
      name: '',
      symbol: '',
      decimals: 0,
      totalSupply: mockBignumber,
    },
  },
  timestamp: 10,
  votingPower: {
    percent: 10,
    userVotingPower: mockBignumber,
    atSnapshot: mockBignumber,
    atCurrentSnapshot: mockBignumber,
  },
  contract: {} as any,
  createTransaction: jest.fn(),
  currentLockedPercent: 10,
};
