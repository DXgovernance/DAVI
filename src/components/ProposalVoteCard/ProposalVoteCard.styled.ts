import { Button } from 'components/Primitives/Button';
import styled, { css } from 'styled-components';

export const ButtonsContainer = styled.div`
  flex-direction: column;
  display: flex;
  margin-top: 2.5rem;
`;

export const SmallButton = styled(Button)`
  margin: 0px;
  padding: 0.1rem 0.4rem;
  width: 60px;
`;

export const VotesContainer = styled.div`
  display: flex;
  flex-direction: column;
  color: ${({ theme }) => theme.colors.text};
`;

export const VoteOptionsLabel = styled.div`
  color: ${({ theme }) => theme.colors.proposalText.grey};
  margin-bottom: 0.5rem;
`;

export const VoteActionButton = styled(Button)`
  height: 2.5rem;

  :disabled {
    background-color: transparent;
    border: 1px solid ${({ theme }) => theme.colors.border.initial};
    color: ${({ theme }) => theme.colors.proposalText.grey};
    opacity: 1;
  }
`;

export const VoteOptionButton = styled(VoteActionButton)<{
  active?: boolean;
  selected?: boolean;
}>`
  margin-bottom: 1rem;
  background-color: ${({ theme }) => theme.colors.muted};

  :active {
    color: ${({ theme }) => theme.colors.background};
    background-color: ${({ theme }) => theme.colors.border.hover};
  }

  ${({ active, selected }) =>
    (active || selected) &&
    css`
      color: ${({ theme }) => theme.colors.background};
      background-color: ${({ theme }) => theme.colors.border.hover};
    `}
`;
