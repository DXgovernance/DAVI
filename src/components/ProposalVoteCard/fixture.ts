import { BigNumber } from 'ethers';
import moment from 'moment';
import { ContractState } from 'types/types.guilds.d';

const mockBignumber = BigNumber.from(100000000);
const mockBigNumberOption = BigNumber.from(100000000000000);
const mockBigNumberTotalSupply = BigNumber.from(1000000000000000);
const mockBigNumberTotalLocked = BigNumber.from(1000000000000000);

export const mockProposalVoteCardProps = {
  proposal: {
    id: '0x0',
    metadata: {
      description: 'This is a test proposal',
      voteOptions: ['Yes', 'No'],
    },
    endTime: moment(),
    creator: '0x1234567890123456789012345678901234567890',
    startTime: moment(),
    timeDetail: '',
    title: 'SWPR single reward campaign',
    to: [],
    value: [],
    totalOptions: BigNumber.from(0),
    contentHash: '0x1234567890123456789012345678901234567890',
    contractState: ContractState.Active,
    totalVotes: [],
  },
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
  timestamp: 10,
  votingPower: {
    percent: 100,
    userVotingPower: mockBignumber,
    atSnapshot: mockBignumber,
    atCurrentSnapshot: mockBignumber,
  },
  contract: {} as any,
  createTransaction: () => {},
  currentLockedPercent: 100,
  userVote: {
    option: '0',
    votingPower: BigNumber.from(100000000),
  },
};

export const mockVoteResults = {
  isPercent: true,
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
  proposalMetadata: {
    description: 'This is a test proposal',
    voteOptions: ['Yes', 'No'],
  },
};

export const mockVoteConfirmationModal = {
  isOpen: true,
  onDismiss: () => {},
  onConfirm: () => {},
  selectedAction: 'Yes',
  votingPower: 100,
  totalLocked: 100,
  onAddToVoteCart: () => {},
};

export const mockVoteChart = {
  isPercent: true,
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
};
