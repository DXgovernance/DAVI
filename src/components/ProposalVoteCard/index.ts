export { default as ProposalVoteCard } from './ProposalVoteCard';
export { confirmVoteProposal } from './utils';
