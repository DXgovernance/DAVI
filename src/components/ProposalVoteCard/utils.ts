import { ConfirmVoteProposalProps, VoteOnProposalProps } from './types';

export const voteOnProposal = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};

export const confirmVoteProposal = ({
  proposal,
  contract,
  selectedOption,
  userVotingPower,
  createTransaction,
  cb,
}: ConfirmVoteProposalProps) => {
  createTransaction(
    `Vote on proposal ${proposal?.title}`,
    async () => contract.setVote(proposal?.id, selectedOption, userVotingPower),
    true,
    cb
  );
};

export const getOptionLabel = ({ metadata, optionKey, t }) => {
  const metadataLabel = metadata?.voteOptions?.[optionKey];
  return metadataLabel
    ? metadataLabel
    : Number(optionKey) === 0
    ? t('against', { defaultValue: 'Against' })
    : t('option', { optionKey });
};
