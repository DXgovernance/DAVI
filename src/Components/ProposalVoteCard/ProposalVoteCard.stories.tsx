import { ComponentStory, ComponentMeta } from '@storybook/react';
import { ProposalVoteCard } from 'Components/ProposalVoteCard';
import { mockProposalVoteCardProps } from './fixture';

export default {
  title: 'ProposalVoteCard',
  component: ProposalVoteCard,
  argTypes: {
    voteData: {
      description:
        'Information about vote options, quorum, total locked and token',
    },
    votingPower: {
      description: 'Information about user voting power',
    },
    proposal: {
      description:
        'Information about proposal id, description, vote options descriptions, and proposal endtime',
    },
    timestamp: {
      description: 'Timestamp of last rerender',
    },
    contract: {
      description: 'ERC20 Guild contract methods',
    },
    currentLockedPercent: {
      description: 'Current locked percent in the contract',
    },
    createTransaction: {
      description: 'Transaction to be created',
    },
  },
} as ComponentMeta<typeof ProposalVoteCard>;

const Template: ComponentStory<typeof ProposalVoteCard> = args => (
  <ProposalVoteCard {...args} />
);

export const Simple = Template.bind({});
Simple.args = mockProposalVoteCardProps;

export const Loading = Template.bind({});
Loading.args = {};
