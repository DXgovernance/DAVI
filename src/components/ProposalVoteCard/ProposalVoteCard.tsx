import {
  SidebarCard,
  SidebarCardContent,
  SidebarCardHeaderSpaced,
} from 'components/SidebarCard';
import VotesChart from './components/VoteChart/VoteChart';
import { VoteConfirmationModal } from './components/VoteConfirmationModal';
import { UserVote } from './components/UserVote';
import VoteResults from './components/VoteResults/VoteResults';
import { BigNumber } from 'ethers';
import moment from 'moment';
import { Loading } from 'components/primitives/Loading';
import { useState, useMemo } from 'react';
import { toast } from 'react-toastify';
import {
  SmallButton,
  VotesContainer,
  ButtonsContainer,
  VoteActionButton,
  VoteOptionButton,
  VoteOptionsLabel,
} from './ProposalVoteCard.styled';
import { voteOnProposal, confirmVoteProposal } from './utils';
import { useTheme } from 'styled-components';
import { hasVotingPowerProps, ProposalVoteCardProps } from './types';
import { useTranslation } from 'react-i18next';
import { getOptionLabel } from 'components/ProposalVoteCard/utils';
import useVotingPowerPercent from 'Modules/Guilds/Hooks/useVotingPowerPercent';

const ProposalVoteCard = ({
  voteData,
  proposal,
  votingPower,
  timestamp,
  contract,
  createTransaction,
  userVote,
}: ProposalVoteCardProps) => {
  const theme = useTheme();
  const { t } = useTranslation();

  const [isPercent, setIsPercent] = useState(true);
  const [selectedAction, setSelectedAction] = useState<BigNumber>();
  const [modalOpen, setModalOpen] = useState<boolean>();
  const isOpen = useMemo(
    () => proposal?.endTime?.isAfter(moment(timestamp)),
    [proposal, timestamp]
  );

  const votedOptionLabel = useMemo(() => {
    if (!userVote?.action) return null;
    return getOptionLabel({
      metadata: proposal?.metadata,
      optionKey: userVote?.action,
      t,
    });
  }, [userVote?.action, proposal?.metadata, t]);

  const toastError = (msg: string) =>
    toast.error(msg, {
      style: {
        backgroundColor: theme.colors.bg1,
        borderColor: theme.colors.border1,
      },
      autoClose: 2800,
      hideProgressBar: true,
    });

  const { hasNoVotingPower, hasVotingPowerAtCurrentSnapshot } = voteOnProposal({
    votingPowerAtProposalSnapshot: votingPower?.atSnapshot,
    votingPowerAtProposalCurrentSnapshot: votingPower?.atCurrentSnapshot,
  });

  const handleVoteOnProposal = ({
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  }: hasVotingPowerProps) => {
    if (hasNoVotingPower) {
      if (hasVotingPowerAtCurrentSnapshot) {
        return toastError(
          'Current voting power gained after proposal creation'
        );
      }
      return toastError('No Voting Power');
    }

    return setModalOpen(true);
  };

  return (
    <SidebarCard
      header={
        <SidebarCardHeaderSpaced>
          {!voteData ? (
            <Loading loading text />
          ) : isOpen ? (
            t('castVote')
          ) : (
            t('voteResults')
          )}
          <SmallButton
            variant="secondary"
            onClick={() => setIsPercent(!isPercent)}
          >
            {!voteData ? (
              <Loading loading text skeletonProps={{ width: 40 }} />
            ) : isPercent ? (
              '%'
            ) : (
              voteData?.token?.symbol
            )}
          </SmallButton>
        </SidebarCardHeaderSpaced>
      }
    >
      <SidebarCardContent>
        <VotesContainer>
          <VoteResults
            isPercent={isPercent}
            voteData={voteData}
            proposalMetadata={proposal?.metadata}
          />
          <VotesChart isPercent={isPercent} voteData={voteData} />
        </VotesContainer>

        <UserVote
          isPercent={isPercent}
          voteData={voteData}
          userVote={userVote}
          votedOptionLabel={votedOptionLabel}
        />
        {/* Hide voting options if user has already voted */}
        {isOpen && !userVote?.action && voteData?.options && (
          <ButtonsContainer>
            <VoteOptionsLabel>{t('options')}</VoteOptionsLabel>

            {/* Getting the full option keys list but displaying default 0 index option at the bottom */}
            {[...Object.keys(voteData?.options).slice(1), '0'].map(
              optionKey => {
                const bItem = BigNumber.from(optionKey);
                const label = getOptionLabel({
                  metadata: proposal?.metadata,
                  optionKey: optionKey,
                  t,
                });

                return (
                  <VoteOptionButton
                    key={optionKey}
                    optionKey={Number(optionKey)}
                    // variant="minimal"
                    active={selectedAction && selectedAction.eq(bItem)}
                    onClick={() => {
                      setSelectedAction(
                        selectedAction && selectedAction.eq(bItem)
                          ? null
                          : bItem
                      );
                    }}
                  >
                    {label}
                  </VoteOptionButton>
                );
              }
            )}

            <VoteActionButton
              disabled={!selectedAction}
              onClick={() =>
                handleVoteOnProposal({
                  hasNoVotingPower,
                  hasVotingPowerAtCurrentSnapshot,
                })
              }
            >
              {t('vote')}
            </VoteActionButton>
          </ButtonsContainer>
        )}
      </SidebarCardContent>

      <VoteConfirmationModal
        isOpen={modalOpen}
        onDismiss={() => setModalOpen(false)}
        onConfirm={() => {
          confirmVoteProposal({
            proposal,
            contract,
            selectedAction,
            userVotingPower: votingPower.userVotingPower,
            createTransaction,
          });
          setModalOpen(false);
          setSelectedAction(null);
        }}
        selectedAction={getOptionLabel({
          metadata: proposal?.metadata,
          optionKey: selectedAction?.toNumber(),
          t,
        })}
        votingPower={votingPower?.percent}
        currentVoteAmount={useVotingPowerPercent(
          voteData?.options?.[selectedAction?.toNumber()],
          voteData?.totalLocked
        )}
      />
    </SidebarCard>
  );
};

export default ProposalVoteCard;
