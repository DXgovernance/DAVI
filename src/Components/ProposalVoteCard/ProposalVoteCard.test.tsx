import { render } from 'utils/tests';
import ProposalVoteCard from './ProposalVoteCard';
import { mockProposalVoteCardProps } from './fixture';

jest.mock('contexts/index', () => jest.fn());

describe('ProposalVoteCard', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <ProposalVoteCard {...mockProposalVoteCardProps} />
    );
    expect(container).toMatchSnapshot();
  });
});
