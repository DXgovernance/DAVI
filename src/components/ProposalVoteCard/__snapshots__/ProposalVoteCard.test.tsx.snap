// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalVoteCard matches the snapshot 1`] = `
.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c1 {
  border: 1px solid #303338;
  border-radius: 10px;
  margin-bottom: 1rem;
  box-shadow: 0px 4px 8px 0px rgba(0,0,0,0.2);
}

.c2 {
  border-bottom: 1px solid #303338;
  padding: 8px 16px;
}

.c3 {
  font-family: 'Inter',-apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 500;
}

.c3 strong,
.c3 b {
  font-weight: 600;
}

.c4 {
  font-weight: 600;
  margin: 0.5rem;
  color: #fff;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  padding: 1rem;
  color: #fff;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10px 0px 40px 0px;
  width: 100%;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 0.75rem;
  border-radius: 32px;
  background-color: #303338;
  overflow: hidden;
}

.c17 {
  width: 10%;
  background: #E75C5C;
  height: 0.75rem;
  overflow: hidden;
}

.c18 {
  width: 10%;
  background: #295FF4;
  height: 0.75rem;
  overflow: hidden;
}

.c20 {
  height: 18px;
  margin-top: 10px;
  width: 1px;
  background: #303338;
}

.c21 {
  padding: 4px 8px;
  border: 1px solid #303338;
  border-radius: 0px 32px 32px;
  font-size: 14px;
  font-weight: 600;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c21 span {
  margin-left: 2px;
}

.c19 {
  width: 65px;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  left: 0%;
}

.c22 {
  margin-right: 0.4rem;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  font-size: 14px;
  border: 1px solid #303338;
  background-color: transparent;
  color: #fff;
  border-radius: 32px;
  padding: 0.5rem 0.8rem;
  margin: 0.2rem;
  width: auto;
}

.c6:disabled {
  color: initial;
  opacity: 0.4;
  cursor: auto;
}

.c6:hover:enabled {
  border-color: #fff;
}

.c6:active:enabled {
  border: 1px solid #303338;
}

.c6:disabled {
  color: #303338;
}

.c23 {
  margin: 1rem 0 0;
  border-radius: 0.375rem;
  padding: 8px;
  font-size: 12px;
  background: #303338;
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #295FF4;
  border-radius: 50%;
}

.c14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #E75C5C;
  border-radius: 50%;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-weight: 500;
  font-size: 14px;
  margin: 5px 0px 5px 0px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c12 {
  margin-right: 0.5rem;
}

.c7 {
  margin: 0px;
  padding: 0.1rem 0.4rem;
  width: 60px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #fff;
}

<div>
   
  <div
    class="c0 c1"
  >
    <div
      class="c0 c2"
    >
      <h2
        class="c3 c4 c5"
      >
        castVote
        <button
          class="c6 c7"
        >
          %
        </button>
      </h2>
    </div>
    <div
      class="c0"
    >
      <div
        class="c0 c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10"
          >
            <div
              class="c11"
            >
              <span
                class="c12"
              >
                <span
                  class="c13"
                  color="#295FF4"
                  size="8"
                />
              </span>
              No
            </div>
            <span>
              10%
            </span>
          </div>
          <div
            class="c10"
          >
            <div
              class="c11"
            >
              <span
                class="c12"
              >
                <span
                  class="c14"
                  color="#E75C5C"
                  size="8"
                />
              </span>
              Yes
            </div>
            <span>
              10%
            </span>
          </div>
          <div
            class="c15"
          >
            <div
              class="c16"
            >
              <div
                class="c17"
                color="#E75C5C"
              />
              <div
                class="c18"
                color="#295FF4"
              />
            </div>
            <div
              class="c19"
            >
              <div
                class="c20"
              />
              <div
                class="c21"
              >
                <svg
                  class="c22"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M243.2 189.9V258c26.1 5.9 49.3 15.6 73.6 22.3v-68.2c-26-5.8-49.4-15.5-73.6-22.2zm223.3-123c-34.3 15.9-76.5 31.9-117 31.9C296 98.8 251.7 64 184.3 64c-25 0-47.3 4.4-68 12 2.8-7.3 4.1-15.2 3.6-23.6C118.1 24 94.8 1.2 66.3 0 34.3-1.3 8 24.3 8 56c0 19 9.5 35.8 24 45.9V488c0 13.3 10.7 24 24 24h16c13.3 0 24-10.7 24-24v-94.4c28.3-12.1 63.6-22.1 114.4-22.1 53.6 0 97.8 34.8 165.2 34.8 48.2 0 86.7-16.3 122.5-40.9 8.7-6 13.8-15.8 13.8-26.4V95.9c.1-23.3-24.2-38.8-45.4-29zM169.6 325.5c-25.8 2.7-50 8.2-73.6 16.6v-70.5c26.2-9.3 47.5-15 73.6-17.4zM464 191c-23.6 9.8-46.3 19.5-73.6 23.9V286c24.8-3.4 51.4-11.8 73.6-26v70.5c-25.1 16.1-48.5 24.7-73.6 27.1V286c-27 3.7-47.9 1.5-73.6-5.6v67.4c-23.9-7.4-47.3-16.7-73.6-21.3V258c-19.7-4.4-40.8-6.8-73.6-3.8v-70c-22.4 3.1-44.6 10.2-73.6 20.9v-70.5c33.2-12.2 50.1-19.8 73.6-22v71.6c27-3.7 48.4-1.3 73.6 5.7v-67.4c23.7 7.4 47.2 16.7 73.6 21.3v68.4c23.7 5.3 47.6 6.9 73.6 2.7V143c27-4.8 52.3-13.6 73.6-22.5z"
                  />
                </svg>
                <span>
                  0
                </span>
                <span>
                  %
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="c0 c23"
        >
          youVotedForOption
           
          with
           
          0%
        </div>
      </div>
    </div>
  </div>
   
</div>
`;
